Generalized Text, Added Subtopic,

florianreihl opened the pull request:
**Images**
Currently only Finance Management has subtopics in the database
<img width="1127" alt="Screenshot 2024-02-01 at 1 10 30 PM" src="https://github.com/aakashg58/PR-DigitalLiteracy/assets/59579743/e57aef0e-7548-4ef4-af0f-cffbf15e2d3e">
After selecting a Subtopic
<img width="1133" alt="Screenshot 2024-02-01 at 1 10 39 PM" src="https://github.com/aakashg58/PR-DigitalLiteracy/assets/59579743/12dca6e5-03fe-4a01-a079-ba5e20a07f28">
Topic with no subtopic
<img width="1114" alt="Screenshot 2024-02-01 at 1 10 56 PM" src="https://github.com/aakashg58/PR-DigitalLiteracy/assets/59579743/790322e4-0e6b-4baa-87f5-56ced3faedd3">

**Changes:**

Intro Sections (ex: FinanceAndManagementIntro)

- Retrieved subtopic: item.subtopic from database entries

Video Pages (ex: TechInDailyLife.jsx)

- Created subtopicValue to hold the current subtopic value
- Created tags to hold the current tags from the search bar
- Created a reset subtopic method used by Breadcrumb
- Layout:
- Search Bar and Breadcrumb are always visible
-  If there are no subtopics or the search bar has tags or a subtopic is selected the YouTubeVideoSection is displayed
- Otherwise, the subtopic selection is displayed

Homepage

- Generalized Text (same as previous pull request)

YouTubeVideoSection

- Added a condition to filter out Subtopic. Videos with no subtopic are not filtered out (otherwise half the videos would be missing).
- Refactored the searchbar into a separate component. I wanted to display the searchbar while the subsection was visible and position the breadcrumb below it.

Breadcrumb

- New component that is positioned below the searchbar to inform the user of their position
- If there are no subtopics then Breadcrumbs is set to video
- Pressing on subtopic resets the current subtopic to []
- Pressing on Home Button returns home

Searchbar

- Refactored out margin box into Video Pages to make the component easier to position with the breadcrumb

Subtopic Selector

- Displays a panels of button for each subsection
- Pressing on a subtopic sets the current subtopic to [subtopic]

IMPROVEMENTS:

- We can generalize the layout of the Video Pages (the Intro Section component is only file specific)
- Intro Sections could be refactored as a component
- YouTubeVideoSection could be refactored as component 
- Move constant values out of files to more easily manage them
- Redundant code in SubtopicSelection and HomeCard. Could extend HomeCard to set subtopic instead of url. 
- Custom Images
- Documentation

Florian-commits committed:
Removed Home Button & Generalized Welcome Text

florianreihl committed:
Merge branch 'aakashg58:main' into main

Florian-commits committed:
Saving Progress

Florian-commits committed:
Saving Changes 2

Florian-commits committed:
Save Changes 3

Florian-commits committed:
Saving Changes

Florian-commits committed:
Refactored Search Bar out of YouTubeVideoSection as a component, implemented subsection selection, implemented breadcrumb

dipunj commented on src/components/Video/Breadcrumb/index.jsx:
```diff
@@ -0,0 +1,34 @@
+			{(subtopics.length == 0 || subtopicValue.length > 0) && (
+				<>
+					<span style={{ margin: '0 5px' }}> {'>'} </span>
+					<div>
+						<b>Videos</b>
+					</div>
```
Instead of the div.b nesting, can you just create a single span, and use CSS class `font-bold` (https://tailwindcss.com/docs/font-weight) to bolden the text. I usually prefer to keep all the styling related aspects with CSS, rather than relying on DOM tags.

dipunj commented on src/components/Video/Breadcrumb/index.jsx:
```diff
@@ -0,0 +1,34 @@
+
+			{subtopicValue.length > 0 ? (
+				<>
+					<span style={{ margin: '0 5px' }}> {'>'} </span>
```
You could just use the utility class `mx-1.5` (https://tailwindcss.com/docs/margin)

dipunj commented on src/pages/TechSafetyAndPrivacy.jsx:
```diff
@@ -47,12 +66,29 @@ function TechSafetyAndPrivacy() {
-			<div className="pl-80">
+
+			<div className="md:pl-80">
 				<TechSafetyAndPrivacyIntro
 					dataFromSafetyAndPrivacyIntro={dataFromSafetyAndPrivacyIntro}
 					dataFromFirebase={dataFromFirebase}
 				/>
-				<YouTubeVideoSection osvalue={osvalue} />
+
+				<Box style={{ margin: 'auto', width: '70%', paddingBottom: '2rem' }}>
+					<Searchbar tagsFromSearchBar={tagsFromSearchBar} tags={tags} />
+					<Breadcrumb subtopicValue={subtopicValue} handleResetSubtopic={handleResetSubtopic} subtopics={subtoptics} />
+				</Box>
+
+				{subtoptics.length == 0 || subtopicValue.length > 0 || tags.length > 0 ? (
+					<YouTubeVideoSection
+						osvalue={osvalue}
+						subtopicValue={subtopicValue}
+						tags={tags}
+					/>
+				) : (
+					<SubtopicSelection
+						dataFromSubtopicSelector={dataFromSubtopicSelector}
+						subtopics={subtoptics} />
+				)}
```
this piece of code seems to be repeated numerous times, are we doing this for easier modifiability? Otherwise I suggest, let's extract it out and avoid code repeat.

dipunj reviewed (changes requested):
Please review my comments, I've requested just a few minor changes.

Florian-commits committed:
Refactored Layout of Video Page and Intro

Florian-commits committed:
Moved json

Florian-commits committed:
fixed name

florianreihl commented:
I refactored all of the redundant code from the video section and intro layouts. I had initially passed the intro component into Techvideos.jsx but felt the code was too redundant as well.

Breadcrumb
1. applied your suggestions

Pages 
1. all video related code was moved to layouts/main/techvideos.jsx
2. unique strings are passed into techvideos.jsx

layouts/main/Techvideos.jsx
1. same code but accepts strings that are passed into components/video/intro

components/video/intro
1. a condensed version of all of the intro pages
2. I kept the .json file around although it doesn't seem anything ever used it
3. Removed filter box

Navbar
1. commented out ask a question

Florian-commits committed:
Deleted redundant files

Florian-commits committed:
Removed Ask a Question & Video Filter Box

Florian-commits committed:
Implemented Logic & refactored intro page

Florian-commits committed:
Made subtopics indepdent from initialpage

florianreihl commented:
https://github.com/aakashg58/PR-DigitalLiteracy/assets/59579743/480f1d61-8b03-4517-bef0-ce148ca8935f


1. I commented out the filter because the apply button was covering up a value. It's still a broken with different screen sizes. Perhaps, the filter panel should be scrollable, but I'd have to spend time figuring out how to separate the scrolling interface depending on what you hover over.

2. I implemented logic for the side panel. Essentially, there are two arrays: one in /techVideos that keeps track of the applied tags and one in /filterPanel that keeps track of selected tags. OnSave() applies the selected tags to 

3. Separated the intro text component from the logic of FirebaseRetrieveVideos. I removed some filtering logic to make it independent from pageValue and the set operating_system. 

- removed code:  `video.operating_system === value && video.category === pageValue`

- I still need to explore how FirebaseRetrieveVideos works.

4. The logic of the subtopic is independent of the initial page. Essentially, if you have 2 content types selected and both have subtopics to display all subtopics (ex: 12 subtopics). If one page has no subtopics, all non-subtopic videos will be displayed/hidden until a subtopic (video subtopic undefined) is selected. If both have no videos just display all videos (subtopic selection undefined).

Let me know if you have any concerns.

Florian-commits committed:
Fixed Bug and Rewrote Sorting Code

Florian-commits committed:
removed a testing line

florianreihl commented:
5. Rewrote the filtering in /youtubevideos to make it easier to understand and implement new filters
6. Removed a bug in /youtubevideos that would use the video.tags as a key and replaced it with the url for now in /FirebaseRetrieveVideos. Hopefully, this fixes the issue Dr. Aakash had with the videos not showing up / conflicting

Florian-commits committed:
changed name of classAndWord to classAndWork

florianreihl commented:
7. Renamed to classAndWord to classAndWork. **NOTE: the database value is still class_word.**

Florian-commits committed:
Aligned the videos better

Florian-commits committed:
Aligned the videos better

florianreihl commented:
8. Centered the videos and subtopics so it looks better with a 10% on both sides. **Important I'm still factoring out the styling of both so they're more identical manage. I temporarily used style=({ width: '100%', height: '300px', background: 'black' })**

Florian-commits committed:
added  margin and fixed tags filter error

Florian-commits committed:
Made the subtopic selection looks similar. Still need to refactor it to make it have the same sizing parameter as the video.

Florian-commits committed:
Added scroll to top effect, reduced footer size, removed redundant foter from login

florianreihl commented:
9. Added scroll to top effect to select pages
10. reduced footer size 
11. removed redundant footer from login

Florian-commits committed:
Replaced subtopic buttons with random colors and added subtopics to list.
